### ISSUE TITLE
Feature Request: normal database queries + parallel queries

### ISSUE THREAD

**Issue by danneu** [denisftw/modern-web-kotlin#1]
The book is great and I would recommend it to other beginners, but it seemed to only contain one database query example, and it was a bit obscured by the fact that it was inside an idiosyncratic vertx provider/middleware.

It's not obvious what the Right Way is to write my own arbitrary queries from here.

For example, imagine a `package db` file that has functions `listUsers`, `insertUser`, `getUser`. Should these return vertx `Future`s or kovenant `Promise`s?

For instance, in a real web app, you often shoot a few queries in parallel and await them all.

In Node.js:

``` javascript
// server.js
router.get('/users', (req, res) => {
  // query the userlist twice in parallel
  Promise.all([db.listUsers(), db.listUsers()]).then(([list1, list2]) => {
    res.send(JSON.stringify([list1, list2])) 
  })
})
```

This was my solution in Kotlin:

``` kotlin
package db

import com.zaxxer.hikari.HikariDataSource
import kotliquery.Row
import kotliquery.queryOf
import kotliquery.sessionOf
import kotliquery.using
import nl.komponents.kovenant.Promise
import nl.komponents.kovenant.deferred

class DbUser(val id: Int, val name: String) {
    companion object {
        fun fromDb(row: Row): DbUser {
            return DbUser(row.int("id"), row.string("name"))
        }
    }
}

fun listUsers (dataSource: HikariDataSource): Promise<List<DbUser>, Exception> {
    val defer = deferred<List<DbUser>, Exception>()
    using(sessionOf(dataSource)) { session ->
        val users: List<DbUser> = session.run(queryOf("""
            SELECT *
            FROM users
        """).map { DbUser.fromDb(it) }.asList)
        defer.resolve(users)
    }
    return defer.promise
}
```

And then inside `MainVerticle.kt`:

``` kotlin
import db

router.get("/users").handler { ctx ->
    val usersP1 = db.listUsers(dataSource)
    val usersP2 = db.listUsers(dataSource)

    all(usersP1, usersP2).success { results ->
        val users1 = results[0]
        val usersP2 = results[1]
        val json = jsonMapper.writeValueAsString(listOf(uses1, users2))
        ctx.response().end(json)
    }
}
```

Am I on the right track / how would you recommend we do this? The mix of async vs sync is new to me coming from ecosystems where either everything is async or everything in sync (one thread per request).

I'm sorry if this isn't the right place. It's a question I could see other people having.

I think your book could answer this question by walking the user through how to implement a `package db` namespace and then implementing a few CRUD endpoints, with maybe one example of composing multiple queries in a parallel fetch. It would immediately help us 1:1 match our experience with other languages with how one might do things with vertx+kotlin.

**Comment by denisftw** [comment]
I'm glad you like the book! 

I agree that using Vert.x Java API is not the best experience ever, but we needed to use their facilities to make other things work. Hopefully, they will make it more Kotlin-friendly in future versions.

As for the `listUsers` function, if you want full control over `Promise` creation, you can totally use the deferred API. In this case, though, you are also expected to catch possible exceptions and `reject` the deferred. On the other hand, it would be much easier to simply wrap everything in `task` and get back a `Promise<List<DbUser>, Exception>`. 

When you need to combine several `Promise`s together, you can certainly use the `all` function. Alternatively, you can also use `bind/map` or `flatMap/map` combinations. I like the latter approach better, because it's more general and works with any monad. Also, as I'm arguing [here](http://appliedscala.com/blog/2016/fp-in-java8-kotlin-scala/), a syntactic construct for writing monad comprehensions in Kotlin actually makes sense and may be added in the future.

Overall, I wouldn't say that querying a database is very different from querying a remote server. In both cases, we're wrapping blocking API calls inside a `Promise` and executing them in a worker thread pool. In our example, this thread pool is maintained by Vert.x, but it's not a big deal. And, of course, you can always lift a value to a `Promise` using the `Promise.of` function.

Probably, I should add these clarifications to the text as well.
